### ISSUE TITLE
[强化学习做市商策略] | RL Market Maker Strategy

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4461]
**标签**: quantitative-strategy
### 摘要
使用强化学习算法优化做市商的最优报价策略，在库存风险与价差收益间取得平衡。
### 算法描述
**核心原理**：
做市商通过提供流动性赚取买卖价差，但面临库存不平衡风险。RL模型学习动态调整报价间距和库存管理。

**关键概念**：
1. **存货风险**：
   Inventory_Risk = E[max(-Q, 0)] * E[max(P_buy - P_sell, 0)]
   其中Q为库存方向（正=多头，负=空头）
2. **价差收益**：
   Spread_Revenue = (Bid - Ask) * min(V_buy, V_sell)
3. **公平价格估计**：
   P_fair = α * P_last + (1-α) * P_mid

**强化学习框架**：
- State: (inventory, spread, volatility, time_to_expiry)
- Action: (bid_offset, ask_offset, size_multiplier)
- Reward: PnL - λ * inventory_risk

**深度Q网络(DQN)**：
```python
class DQNetwork(nn.Module):
    def __init__(self, state_dim, action_dim):
        self.fc1 = nn.Linear(state_dim, 128)
        self.fc2 = nn.Linear(128, 64)
        self.fc3 = nn.Linear(64, action_dim)
```

**epsilon-greedy策略**：
- 前期高探索（epsilon=1.0）
- 后期高利用（epsilon=0.05）
- epsilon_decay = 0.995
### Python代码示例
```python
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from collections import deque, namedtuple
import random
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
from datetime import datetime
from enum import Enum

# Experience replay buffer
Experience = namedtuple('Experience', 
    ['state', 'action', 'reward', 'next_state', 'done'])

@dataclass
class OrderBookLevel:
    price: float
    size: float

@dataclass
class MarketState:
    timestamp: datetime
    best_bid: float
    best_ask: float
    mid_price: float
    inventory: float
    volatility: float
    time_to_expiry: float

@dataclass
class Quote:
    bid_price: float
    bid_size: float
    ask_price: float
    ask_size: float
    timestamp: datetime

class DQNetwork(nn.Module):
    """深度Q网络"""
    def __init__(self, state_dim: int, action_dim: int, hidden_dim: int = 128):
        super().__init__()
        self.fc1 = nn.Linear(state_dim, hidden_dim)
        self.fc2 = nn.Linear(hidden_dim, hidden_dim // 2)
        self.fc3 = nn.Linear(hidden_dim // 2, action_dim)
        self.dropout = nn.Dropout(0.2)
    
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = torch.relu(self.fc1(x))
        x = self.dropout(x)
        x = torch.relu(self.fc2(x))
        x = self.fc3(x)
        return x

class ReplayBuffer:
    """经验回放缓冲区"""
    def __init__(self, capacity: int = 10000):
        self.buffer = deque(maxlen=capacity)
    
    def push(self, state, action, reward, next_state, done):
        self.buffer.append(Experience(state, action, reward, next_state, done))
    
    def sample(self, batch_size: int) -> List[Experience]:
        return random.sample(self.buffer, batch_size)
    
    def __len__(self):
        return len(self.buffer)

class RLMarketMaker:
    """强化学习做市商"""
    def __init__(
        self,
        state_dim: int = 6,
        action_dim: int = 9,
        learning_rate: float = 0.001,
        gamma: float = 0.99,
        epsilon: float = 1.0,
        epsilon_decay: float = 0.995,
        epsilon_min: float = 0.05,
        batch_size: int = 32,
        target_update_freq: int = 10
    ):
        self.state_dim = state_dim
        self.action_dim = action_dim
        self.gamma = gamma
        self.epsilon = epsilon
        self.epsilon_decay = epsilon_decay
        self.epsilon_min = epsilon_min
        self.batch_size = batch_size
        self.target_update_freq = target_update_freq
        
        # Q网络和目标网络
        self.q_network = DQNetwork(state_dim, action_dim)
        self.target_network = DQNetwork(state_dim, action_dim)
        self.target_network.load_state_dict(self.q_network.state_dict())
        
        self.optimizer = optim.Adam(self.q_network.parameters(), lr=learning_rate)
        self.replay_buffer = ReplayBuffer(capacity=10000)
        
        self.steps = 0
        self.training_history = []
        
        # 价格偏移选项
        self.offset_options = [0.01, 0.02, 0.03, 0.05, 0.07, 0.10, 0.15, 0.20, 0.30]
        
    def state_to_tensor(self, state: MarketState) -> torch.Tensor:
        """将市场状态转换为张量"""
        return torch.FloatTensor([
            state.inventory / 1000,
            state.mid_price / 100,
            np.log(state.volatility + 1e-8),
            state.time_to_expiry / 86400,
            state.best_bid / state.mid_price,
            state.best_ask / state.mid_price
        ])
    
    def get_action(self, state: MarketState, training: bool = True) -> Tuple[int, float]:
        """epsilon-greedy策略选择动作"""
        if training and random.random() < self.epsilon:
            action_idx = random.randint(0, self.action_dim - 1)
        else:
            with torch.no_grad():
                state_tensor = self.state_to_tensor(state).unsqueeze(0)
                q_values = self.q_network(state_tensor)
                action_idx = q_values.argmax().item()
        
        return action_idx, self.offset_options[action_idx]
    
    def calculate_spread(self, offset: float, mid_price: float) -> Tuple[float, float]:
        """计算买卖报价"""
        half_spread = mid_price * offset
        bid = mid_price - half_spread
        ask = mid_price + half_spread
        return bid, ask
    
    def calculate_reward(
        self,
        trade_price: float,
        mid_price: float,
        inventory: float,
        inventory_cost: float = 0.001
    ) -> float:
        """计算奖励"""
        # 价差收益
        spread_pnl = abs(trade_price - mid_price)
        
        # 库存成本（惩罚）
        inventory_penalty = inventory_cost * abs(inventory)
        
        return spread_pnl - inventory_penalty
    
    def update_network(self):
        """更新Q网络"""
        if len(self.replay_buffer) < self.batch_size:
            return
        
        batch = self.replay_buffer.sample(self.batch_size)
        states = torch.stack([self.state_to_tensor(e.state) for e in batch])
        actions = torch.LongTensor([e.action for e in batch])
        rewards = torch.FloatTensor([e.reward for e in batch])
        next_states = torch.stack([self.state_to_tensor(e.next_state) for e in batch])
        dones = torch.FloatTensor([e.done for e in batch])
        
        # 当前Q值
        q_values = self.q_network(states).gather(1, actions.unsqueeze(1)).squeeze()
        
        # 目标Q值
        with torch.no_grad():
            next_q_values = self.target_network(next_states).max(1)[0]
            target_q = rewards + self.gamma * next_q_values * (1 - dones)
        
        # 损失和优化
        loss = nn.MSELoss()(q_values, target_q)
        self.optimizer.zero_grad()
        loss.backward()
        torch.nn.utils.clip_grad_norm_(self.q_network.parameters(), 1.0)
        self.optimizer.step()
        
        # 更新目标网络
        if self.steps % self.target_update_freq == 0:
            self.target_network.load_state_dict(self.q_network.state_dict())
        
        # epsilon衰减
        self.epsilon = max(self.epsilon_min, self.epsilon * self.epsilon_decay)
        
        self.steps += 1
        return loss.item()
    
    def generate_quote(
        self,
        state: MarketState,
        base_size: float = 1.0,
        training: bool = True
    ) -> Quote:
        """生成报价"""
        action_idx, offset = self.get_action(state, training)
        
        bid_price, ask_price = self.calculate_spread(offset, state.mid_price)
        
        # 根据动作调整数量
        size_multiplier = 0.5 + (action_idx % 3) * 0.25
        size = base_size * size_multiplier
        
        return Quote(
            bid_price=bid_price,
            bid_size=size,
            ask_price=ask_price,
            ask_size=size,
            timestamp=state.timestamp
        )

class MarketSimulator:
    """市场模拟器"""
    def __init__(self, initial_price: float = 100.0, volatility: float = 0.02):
        self.price = initial_price
        self.volatility = volatility
        self.inventory = 0.0
        self.trade_history = []
    
    def step(self, bid: float, ask: float, size_bid: float, size_ask: float) -> Tuple[float, bool, bool]:
        """模拟市场"""
        # 价格随机游走
        dt = 1/252
        dW = np.random.randn() * np.sqrt(dt)
        self.price *= np.exp(self.volatility * dW)
        
        mid = self.price
        bought = False
        sold = False
        
        # 成交判断（简化）
        if np.random.random() < 0.3:
            if np.random.random() < 0.5 and bid >= mid:
                self.inventory += size_bid
                bought = True
            elif np.random.random() < 0.5 and ask <= mid:
                self.inventory -= size_ask
                sold = True
        
        trade_price = bid if bought else (ask if sold else mid)
        return trade_price, bought, sold
    
    def get_state(self, timestamp: datetime, time_to_expiry: float = 86400) -> MarketState:
        """获取当前市场状态"""
        return MarketState(
            timestamp=timestamp,
            best_bid=self.price * 0.999,
            best_ask=self.price * 1.001,
            mid_price=self.price,
            inventory=self.inventory,
            volatility=self.volatility,
            time_to_expiry=time_to_expiry
        )

def train_rl_market_maker(
    episodes: int = 100,
    max_steps: int = 100,
    initial_price: float = 100.0
) -> Dict:
    """训练RL做市商"""
    agent = RLMarketMaker(state_dim=6, action_dim=9)
    simulator = MarketSimulator(initial_price=initial_price)
    
    episode_rewards = []
    
    for episode in range(episodes):
        simulator = MarketSimulator(initial_price=initial_price)
        total_reward = 0
        
        for step in range(max_steps):
            state = simulator.get_state(datetime.now())
            
            # 生成报价
            quote = agent.generate_quote(state)
            
            # 模拟交易
            trade_price, bought, sold = simulator.step(
                quote.bid_price, quote.ask_price,
                quote.bid_size, quote.ask_size
            )
            
            # 计算奖励
            reward = agent.calculate_reward(
                trade_price, state.mid_price, state.inventory
            )
            
            # 获取下一状态
            next_state = simulator.get_state(datetime.now())
            
            # 存储经验
            agent.replay_buffer.push(state, 0, reward, next_state, False)
            
            # 更新网络
            loss = agent.update_network()
            
            total_reward += reward
        
        episode_rewards.append(total_reward)
        
        if episode % 10 == 0:
            print(f"Episode {episode}: Total Reward = {total_reward:.2f}, Epsilon = {agent.epsilon:.3f}")
    
    return {
        "episode_rewards": episode_rewards,
        "final_epsilon": agent.epsilon
    }

# 测试
if __name__ == "__main__":
    result = train_rl_market_maker(episodes=50, max_steps=50)
    print(f"Training completed. Final epsilon: {result['final_epsilon']:.3f}")
